src/nvim/change.c

```diff
@@ -0,0 +1,1805 @@
+// This is an open source non-commercial project. Dear PVS-Studio, please check
+// it. PVS-Studio Static Code Analyzer for C, C++ and C#: http://www.viva64.com
+
+/// change.c: functions related to changing text
+
+#include "nvim/assert.h"
+#include "nvim/buffer.h"
+#include "nvim/buffer_updates.h"
+#include "nvim/change.h"
+#include "nvim/charset.h"
+#include "nvim/cursor.h"
+#include "nvim/diff.h"
+#include "nvim/edit.h"
+#include "nvim/eval.h"
+#include "nvim/fileio.h"
+#include "nvim/fold.h"
+#include "nvim/indent.h"
+#include "nvim/indent_c.h"
+#include "nvim/mark.h"
+#include "nvim/memline.h"
+#include "nvim/misc1.h"
+#include "nvim/move.h"
+#include "nvim/option.h"
+#include "nvim/screen.h"
+#include "nvim/search.h"
+#include "nvim/state.h"
+#include "nvim/ui.h"
+#include "nvim/undo.h"
+
+#ifdef INCLUDE_GENERATED_DECLARATIONS
+# include "change.c.generated.h"
+#endif
+
+/// If the file is readonly, give a warning message with the first change.
+/// Don't do this for autocommands.
+/// Doesn't use emsg(), because it flushes the macro buffer.
+/// If we have undone all changes b_changed will be false, but "b_did_warn"
+/// will be true.
+/// "col" is the column for the message; non-zero when in insert mode and
+/// 'showmode' is on.
+/// Careful: may trigger autocommands that reload the buffer.
+void change_warning(int col)
+{
+  static char *w_readonly = N_("W10: Warning: Changing a readonly file");
+
+  if (curbuf->b_did_warn == false
+      && curbufIsChanged() == 0
+      && !autocmd_busy
+      && curbuf->b_p_ro) {
+    curbuf_lock++;
+    apply_autocmds(EVENT_FILECHANGEDRO, NULL, NULL, false, curbuf);
+    curbuf_lock--;
+    if (!curbuf->b_p_ro) {
+        return;
+    }
+    // Do what msg() does, but with a column offset if the warning should
+    // be after the mode message.
+    msg_start();
+    if (msg_row == Rows - 1) {
+        msg_col = col;
+    }
+    msg_source(HL_ATTR(HLF_W));
+    msg_ext_set_kind("wmsg");
+    MSG_PUTS_ATTR(_(w_readonly), HL_ATTR(HLF_W) | MSG_HIST);
+    set_vim_var_string(VV_WARNINGMSG, _(w_readonly), -1);
+    msg_clr_eos();
+    (void)msg_end();
+    if (msg_silent == 0 && !silent_mode && ui_active()) {
+      ui_flush();
+      os_delay(1000L, true);  // give the user time to think about it
+    }
+    curbuf->b_did_warn = true;
+    redraw_cmdline = false;  // don't redraw and erase the message
+    if (msg_row < Rows - 1) {
+        showmode();
+    }
+  }
+}
+
+/// Call this function when something in the current buffer is changed.
+///
+/// Most often called through changed_bytes() and changed_lines(), which also
+/// mark the area of the display to be redrawn.
+///
+/// Careful: may trigger autocommands that reload the buffer.
+void changed(void)
```

Comment:
This will cause "readability/fn_size" with linting still.
Can it be explicitly ignored, e.g. with an inline comment?